### PR TITLE
cmd/strelaysrv: Add '-max-connections', '-limit-interval' and '-unlimited-devices' flag

### PR DESCRIPTION
**Pull request by XeroAlpha**
I noticed that the number of connections significantly affects the latency on some relay servers with poor performance. On my own simple application server, about a few thousand connections can cause a TLS handshake to take more than 20 seconds, while syncthing only allows [10 seconds of timeout](https://github.com/syncthing/syncthing/blob/716b42103a7296c6a5ebcb0886c43666d2278ae4/lib/connections/service.go#L70), directly causing each session to terminate with "Listen (BEP/relay): TLS handshake: i/o timeout".

This pull request limits the maximum number of connections for the relay server, and allows setting the interval for checking whether the number of connections is exceeded, and the device list that is exempt from the limit.

### CODE REVIEW

**Comment by bt90** [comment]
I'm not arguing whether these are useful additions but the problem you're describing sounds like the server is simply running out of entropy?

**Comment by XeroAlpha** [comment]
> I'm not arguing whether these are useful additions but the problem you're describing sounds like the server is simply running out of entropy?

In my scenario, I first tried to use the command line `strelaysrv -global-rate=300000` to start the server, and then found that at a specific time (usually local time (GMT+8) 0 o’clock), the actual network rate of the server suddenly increased to several times larger than the set value, and I could not connect to the relay server, but restarting the relay server make everything return to normal. Then I added the `-per-session-rate=3000` flag, but after the specific time, the actual network rate increased again, and there was a large difference between the statistic on `/status` and actual network rate, and my syncthing client always prompted `Listen (BEP/relay): TLS handshake: i/o timeout` when connecting to the server. At this time I used [testutil](https://github.com/syncthing/syncthing/tree/main/cmd/strelaysrv/testutil) to test, found that can quickly and normally join relay and receive invitation, but proxying data from one-side to other will take more than 10 seconds, restarting the relay server will shortened this timeout to less than 0.2 seconds. I will continue to investigate.

**Comment by bt90** [comment]
Could you install `rng-tools` on your system and check if this improves the situation?

You can monitor the available entropy of the kernel:

    cat /proc/sys/kernel/random/entropy_avail

Note that the interpretation of the value changed with Linux kernel 5.10.119 and newer. It is now capped at 256.

**Comment by XeroAlpha** [comment]
```
[root@sas ~]# cat /proc/sys/kernel/random/entropy_avail
3780
```

I installed `rng-tools` yesterday and found that the same situation as before happened at 0 o’clock today (GMT+8). I tried to write a [monitor](https://github.com/XeroAlpha/strelaysrv-monitor) that reads the status once a minute and simulates a handshake to test delay. I found that `numPendingSessionKeys` surged, which may be the reason for the inability to connect.

establish_time(ms): server and client register to relay and start to proxy
handshake_time(ms): server and client send 128 random bytes to each other

|time (GMT+8)|kbps_1m|num_active_sessions|num_connections|num_pending_session_keys|num_proxies|establish_time|handshake_time|
|---|---|---|---|---|---|---|---|
|2023-05-24 23:50:13|806|38|781|0|68|4221|7382|
|2023-05-24 23:51:09|798|40|953|0|73|154|7402|
|2023-05-24 23:52:11|802|39|960|0|74|1395|7634|
|2023-05-24 23:53:08|787|41|957|0|75|160|6480|
|2023-05-24 23:54:08|804|43|956|0|77|158|6779|
|2023-05-24 23:55:10|808|40|954|1|70|163|8506|
|2023-05-24 23:56:09|791|38|951|2|66|411|7330|
|2023-05-24 23:57:11|798|41|946|0|72|2413|7188|
|2023-05-24 23:58:14|809|44|937|1|80|2532|9085|
|2023-05-24 23:59:10|794|41|935|0|77|1058|7723|
|2023-05-25 00:00:08|797|37|940|3|70|150|6833|
|2023-05-25 00:01:08|787|39|939|0|73|147|6126|
|2023-05-25 00:02:09|819|42|925|0|81|162|7861|
|2023-05-25 00:03:07|785|41|983|0|78|157|4991|
|2023-05-25 00:04:10|819|45|990|0|82|156|8873|
|2023-05-25 00:05:08|782|44|997|1|80|166|6399|
|2023-05-25 00:06:12|834|50|1004|4|93|147|9784|
|2023-05-25 00:07:10|778|59|1008|45|105|172|8026|
|2023-05-25 00:08:08|800|78|1003|291|133|161|7099|
|2023-05-25 00:09:10|798|78|995|791|130|162|8269|
|2023-05-25 00:10:10|803|93|989|1456|149|154|7905|
|2023-05-25 00:11:12|806|91|987|1772|157|169|10798|
|2023-05-25 00:12:16|823|112|1002|1754|179|179|13513|
|2023-05-25 00:13:15|803|101|999|1519|166|160|13821|
|2023-05-25 00:14:14|789|101|1004|1069|170|166|11746|
|2023-05-25 00:15:11|787|100|1000|749|159|157|9602|
|2023-05-25 00:16:11|800|97|1004|698|155|163|9379|
|2023-05-25 00:17:10|792|105|1005|809|176|171|8702|
|2023-05-25 00:18:11|799|109|1003|1110|181|208|9048|

**Comment by AudriusButkevicius** [comment]
I guess you can enable logging. I suspect there is something (external), DDoSing the relay at midnight? Relays are quite generic and anyone can usethem (it's not syncthing specific, the protocol can be used for anything), perhaps they are being misused by someone?

File: cmd/strelaysrv/main.go [hunk 1]
```diff
@@ -306,15 +327,18 @@ func main() {
...
-		} else if overLimit.CompareAndSwap(true, false) {
-			log.Println("Dropped below our connection limits. Accepting new connections.")
+			if !overLimit.Swap(true) {
+				log.Println("Gone past our connection limits. Starting to refuse new/drop idle connections.")
+			}
+		} else {
+			if overLimit.CompareAndSwap(true, false) {
+				log.Println("Dropped below our connection limits. Accepting new connections.")
+			}
+			limitCheckTimer.Reset(limitCheckInterval)
```
Comment (AudriusButkevicius) [inline cmd/strelaysrv/main.go:340]: Why is this moving into the else clause? 
Given this is a timer, not a ticker, C fires only once (or once after reset)

File: cmd/strelaysrv/status.go [hunk 2]
```diff
@@ -73,6 +74,7 @@ func getStatus(w http.ResponseWriter, _ *http.Request) {
 		"message-timeout":  messageTimeout / time.Second,
 		"per-session-rate": sessionLimitBps,
 		"global-rate":      globalLimitBps,
+		"max-connection":   descriptorLimit,
```
Comment (AudriusButkevicius) [inline cmd/strelaysrv/status.go:77]: plural?

File: cmd/strelaysrv/main.go [hunk 3]
```diff
@@ -140,15 +147,18 @@ func main() {
 
 	log.Println(longVer)
 
-	maxDescriptors, err := osutil.MaximizeOpenFileLimit()
-	if maxDescriptors > 0 {
-		// Assume that 20% of FD's are leaked/unaccounted for.
-		descriptorLimit = int64(maxDescriptors*80) / 100
-		log.Println("Connection limit", descriptorLimit)
-
+	if descriptorLimit == 0 {
```
Comment (AudriusButkevicius) [inline cmd/strelaysrv/main.go:150]: this should really be called connectionLimit

**Comment by XeroAlpha** [comment]
RelayFull will trigger all connecting client to switch to another relay server. I will create a new pull request later to add a soft connection limit which will only deny newly coming clients when connection count goes up to the limit.

**Pull request closed without merging**